### PR TITLE
[Improve][Connector-v2] Use checkpointId as the commit's identifier instead of the hash for streaming write of paimon sink

### PR DESCRIPTION
**Pull request by dailai**
### Check list

* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update [plugin-mapping.properties](https://github.com/apache/seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  2. Update the pom file of [seatunnel-dist](https://github.com/apache/seatunnel/blob/dev/seatunnel-dist/pom.xml)
  3. Add ci label in [label-scope-conf](https://github.com/apache/seatunnel/blob/dev/.github/workflows/labeler/label-scope-conf.yml)
  4. Add e2e testcase in [seatunnel-e2e](https://github.com/apache/seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e/)
  5. Update connector [plugin_config](https://github.com/apache/seatunnel/blob/dev/config/plugin_config)
* [ ] Update the [`release-note`](https://github.com/apache/seatunnel/blob/dev/release-note.md).

### CODE REVIEW

**Comment by Hisoka-X** [comment]
cc @hawk9821 too

File: seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/commit/PaimonAggregatedCommitter.java [hunk 1]
```diff
@@ -99,8 +111,14 @@ public List<PaimonAggregatedCommitInfo> commit(
 
     @Override
     public PaimonAggregatedCommitInfo combine(List<PaimonCommitInfo> commitInfos) {
-        List<List<CommitMessage>> committables = new ArrayList<>();
-        commitInfos.forEach(commitInfo -> committables.add(commitInfo.getCommittables()));
+        Map<Long, List<CommitMessage>> committables = new HashMap<>();
+        commitInfos.forEach(
```
Comment (hawk9821) [inline seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/commit/PaimonAggregatedCommitter.java:115]: suggest using  stream api groupby

Comment (dailai) [reply seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/commit/PaimonAggregatedCommitter.java:115]: Thinks. Comparing the two implementations, this one is currently more readable, so I chose this one

**Review by hawk9821 — APPROVED** [review]
LGTM  thanks

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/SinkWriter.java [hunk 2]
```diff
@@ -64,6 +64,20 @@ default void applySchemaChange(SchemaChangeEvent event) throws IOException {}
...
+     * need to use 2pc, you can return the commit info in this method, and receive the commit info
+     * in {@link SinkCommitter#commit(List)}. If this method failed (by throw exception), **Only**
+     * Spark engine will call {@link #abortPrepare()}
+     *
+     * @param checkpointId checkpointId
+     * @return the commit info need to commit
+     * @throws IOException If fail to prepareCommit
+     */
+    default Optional<CommitInfoT> prepareCommit(long checkpointId) throws IOException {
+        return Optional.empty();
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/SinkWriter.java]:
```suggestion
        return prepareCommit();
```

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java [hunk 3]
```diff
@@ -238,7 +245,14 @@ public Optional<MultiTableCommitInfo> prepareCommit() throws IOException {
                                                             .entrySet()) {
                                         Optional<?> commit;
                                         try {
-                                            commit = sinkWriterEntry.getValue().prepareCommit();
+                                            SinkWriter<SeaTunnelRow, ?, ?> sinkWriter =
```
Comment (Hisoka-X) [inline seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java:246]: how about change `prepareCommit` to `prepareCommit(checkpointId)`? So we can avoid check `SupportCheckpointIdDownStream`.

Comment (dailai) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java:246]: I thought so at first, but there were so many writers to change, Also schema evolution here (applySchemaChange(SchemaChangeEvent event)) also calls prepareCommit() and has no way to pass checkpointId.

截屏2024-10-15 10 29 47


截屏2024-10-15 10 31 24

Comment (Hisoka-X) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java:246]: > I thought so at first, but there were so many writers to change

After update https://github.com/apache/seatunnel/pull/7835#discussion_r1800327463, I think the change should be less.

> Also schema evolution here (applySchemaChange(SchemaChangeEvent event)) also calls prepareCommit()

Looks like it is problem of JDBCWriter, not engine. The jdbc sink writer does not care checkpointid when invoke `prepareCommit`.

**Comment by Hisoka-X** [comment]
> Use checkpointId as the commit's identifier instead of the hash for streaming write of paimon sink

Is there any benefit in doing this?

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java (hunk 3 above)
Comment (dailai) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java:246]: Yes, you mean if change it directly to prepare(checkpointId), can the jdbc writer pass an  any id value here?

**Comment by dailai** [comment]
> > Use checkpointId as the commit's identifier instead of the hash for streaming write of paimon sink
> 
> Is there any benefit in doing this?

When writing the partition table, paimon expects this identifier to be incremented.It can trigger the historical partition writer shutdown logic here in the paimon source code.

image

File: seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java (hunk 3 above)
Comment (Hisoka-X) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java:246]: > can the jdbc writer pass an any id value here?

This is fine for now.

Comment (dailai) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java:246]: Ok, Thinks @Hisoka-X . I will do it.

Comment (Hisoka-X) [reply seatunnel-api/src/main/java/org/apache/seatunnel/api/sink/multitablesink/MultiTableSinkWriter.java:246]: cc @hailin0

File: seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/PaimonSinkWriter.java [hunk 4]
```diff
@@ -174,16 +175,21 @@ public void write(SeaTunnelRow element) throws IOException {
...
+    @Override
+    public Optional<PaimonCommitInfo> prepareCommit(long checkpointId) throws IOException {
         try {
             List<CommitMessage> fileCommittables;
             if (JobContextUtil.isBatchJob(jobContext)) {
                 fileCommittables = ((BatchTableWrite) tableWrite).prepareCommit();
             } else {
                 fileCommittables =
-                        ((StreamTableWrite) tableWrite).prepareCommit(false, committables.size());
+                        ((StreamTableWrite) tableWrite).prepareCommit(false, checkpointId);
```
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/PaimonSinkWriter.java:189]: Is `committables.size()` to `checkpointId` right?

Comment (dailai) [reply seatunnel-connectors-v2/connector-paimon/src/main/java/org/apache/seatunnel/connectors/seatunnel/paimon/sink/PaimonSinkWriter.java:189]: It's right. The original is always 0. This is the official example of paimon.

截屏2024-10-15 16 32 59

**Review by Hisoka-X — APPROVED** [review]

**Review by Carl-Zhou-CN — APPROVED** [review]
+1

**Pull request merged**